### ISSUE TITLE
Problems with COW & non-COW filesystems

### ISSUE THREAD

**Issue by PiotrZSL** [SiaFoundation/hostd#367]
### Current Behavior

Currently hostd allocate disk space as one big file, split it into sectors and provide to renter.
Volumes are fully pre-allocated, and initialized with random data.

This causes issues on COW filesystems (like ZFS) when lot of disk space is allocated to hostd:
- Low write latency - because most of disk is filled already by random data, when new data need to be stored, free inode need to be found. Searching for free inode takes much longer.
- Higher fragmentation.
- Higher disk wear - simply because only small part of disk is utilized because rest of disk is holding random data

On non-COW file systems:
- Higher disk wear - function 'emptyLocationInVolume' going to return basically first free sectors, this mean that some sectors going to be constantly re-used, and parts at the end of hard-disk won't be ever used unless volume gets full.

### Expected Behavior

It's fine if hostd create volume as one big file, but disk space under this fille should not be filled with random data (that's just malicious). If data would be filled with zeros, then at-least file-system based compression would kick in.

Proper solution would be fallocate file instead of filing it with anything, and when releasing sectors additionally punch-a-hole in file.

### Version

v1.0.4

### What operating system did the problem occur on (e.g. Ubuntu 22.04, macOS 12.0, Windows 11)?

Ubuntu 23.10

### Anything else?

https://github.com/torvalds/linux/blob/ed30a4a51bb196781c8058073ea720133a65596f/include/uapi/linux/falloc.h#L6
https://www.man7.org/linux/man-pages/man2/fallocate.2.html
https://github.com/detailyang/go-fallocate

```
import (
	"os"
	"syscall"
)

func allocateSectors(file *os.File, offset int64, length int64) error {
	if length == 0 {
		return nil
	}
	return syscall.Fallocate(int(file.Fd()), 0, offset, length)
}

func clearSectors(file *os.File, offset int64, length int64) error {
	if length == 0 {
		return nil
	}
	return syscall.Fallocate(int(file.Fd()), 3, offset, length)
}
```

**Comment by PiotrZSL** [comment]
Also adding new volume would be very fast.

**Comment by PiotrZSL** [comment]
As for linux/windows, those operating systems got own api, you still deliver different binaries, making use of different api on those should be simply a compilation specific. Windows got [FSCTL_SET_ZERO_DATA](https://learn.microsoft.com/en-us/windows/win32/api/winioctl/ni-winioctl-fsctl_set_zero_data).

As for database, solution is easy. Put additional column numberOfWrites, increment it every time sector is used, and just put index on it to make a sort. Should be sufficient to utilize all sectors.

Still best would be to simply fallocate file / release memory in middle.

**Comment by n8maninger** [comment]
Fixed by #369
